File: leetcode3/MinimumChangestoMakeKSemipalindromes/a.cpp
/* Copyright (C) 2023 by iamslash */

#include <cstdio>
#include <string>

using namespace std;

// 72ms 99.15% 6.9MB 93.72%
// recursive dynamic programming
// O(N^3) (N)
// difficult
class Solution {
private:
  // [index of s][group count] = minimum ops
  int finalMinOps[201][101] = {};
  // [index of s][length of substring] = minimum ops
  int semiMinOps[201][201];
  int divMinOps[201][201];
  
  int dfs(const string& s, int k, int i) {
    int n = s.size();
    const int MAX_CNT = 10000;
    // base
    if (i == n && k == 0) {
      return 0;
    }
    if (i >= n || k <= 0) {
      return MAX_CNT;
    }
    // memo
    if (finalMinOps[i][k] > 0) {
      return finalMinOps[i][k] - 1;
    }
    // recursion
    finalMinOps[i][k] = MAX_CNT;
    for (int len = 2; i + len <= n; ++len) {
      finalMinOps[i][k] = min(finalMinOps[i][k],
                              1 + semiMinOps[i][len] + dfs(s, k - 1, i + len));
    }
    return finalMinOps[i][k] - 1;
  }
public:
  int minimumChanges(string s, int k) {
    int n = s.size();
    memset(semiMinOps, 1, sizeof(semiMinOps));
    for (int d = 1; d <= n / 2; ++d) {
      memset(divMinOps, 0, sizeof(divMinOps));
      for (int len = 2 * d; len <= n; len += d) {
        for (int i = 0; i + len <= n; ++i) {
          int ops = 0;
          for (int off = 0; off < d; ++off) {
            ops += s[i + off] != s[i + len - (d - off)];
          }
          divMinOps[i][len] = ops + divMinOps[i + d][len - 2 * d];
          semiMinOps[i][len] = min(semiMinOps[i][len], divMinOps[i][len]);
        }
      }
    }
    return dfs(s, k, 0);
  }
};

int main() {
  return 0;
}

File: leetcode3/MinimumChangestoMakeKSemipalindromes/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2023 by iamslash

from collections import defaultdict
from functools import cache

# 686ms 80.12% 29.1MB 56.78%
# recursive dynamic programming
# diffcult
class Solution:
    def minimumChanges(self, s: str, k: int) -> int:
        div = defaultdict(lambda: [1])
        n = len(s)
        for d in range(2, n):
            for v in range(d + d, n + 1, d):
                div[v].append(d)

        def change(i, j):
            return min(semi(i, j, d) for d in div[j - i])

        @cache
        def semi(i, j, d):
            if i >= j:
                return 0
            return semi(i + d, j - d, d) + sum(s[i + ii] != s[j - d + ii] for ii in range(d))

        @cache
        def dp(i, k):
            if k == 1:
                return change(0, i)
            return min(dp(j, k - 1) + change(j, i) for j in range((k - 1) * 2, i - 1))

        return dp(n, k)

if __name__ == "__main__":
    sln = Solution()
    print(sln.minimumChanges("abcab", 2))
